Flow: Pathlines seem to never stop advecting

![flow](https://user-images.githubusercontent.com/9522770/216748337-0cb6642a-f09c-4c88-8f18-ecf3af707035.gif)

Pathlines seem to get "stuck" during their advection, leading to increasing rendering times as more pathlines are injected into the scene.

To reproduce:
1. Download the attached .zip file that contains a BOV files, a binary data file, and a session file.
2. Load the session file and point it to the attached data.  Failing that, load the data and create a flow renderer as follows:
    - Pathlines
    - X, Y, and Z Field Variables = a
    - Pathline Length = 1
    - Injection Interval = 1
    - X, Y, and Z axis seeds = 10
5. Enable image capture and run the time series

The images on my computer and Casper are generated quickly at first but then slow down.  It only takes 5 time steps for the pathlines to completely exit the domain.  Rendering time should theoretically plateau after 5 time steps.

Many renderings captured in the first minute at the beginning of the time series:
![Screen Shot 2023-02-02 at 11 04 00 AM](https://user-images.githubusercontent.com/9522770/216406485-69641340-67c7-4dc3-9e5b-ed8ae1fcfbac.png)

One capture per 2 minutes near timestep 100:
![Screen Shot 2023-02-02 at 11 04 53 AM](https://user-images.githubusercontent.com/9522770/216406793-751d2380-cc71-4362-b666-fcea57434256.png)


[flowBug3325.zip](https://github.com/NCAR/VAPOR/files/10572883/flowBug3325.zip)

@sgpearse  Thanks for the instructions, but I still don't have luck building it on Casper. Following your steps, I have a linker error at the last linking step, thus unable to build VAPOR on Casper.
```
[100%] Linking CXX executable ../../bin/vapor
../../lib/librender.so: undefined reference to `TIFFReadDirectory@LIBTIFF_4.0'
../../lib/librender.so: undefined reference to `_TIFFfree@LIBTIFF_4.0'
../../lib/librender.so: undefined reference to `TIFFGetField@LIBTIFF_4.0'
../../lib/librender.so: undefined reference to `TIFFScanlineSize@LIBTIFF_4.0'
../../lib/librender.so: undefined reference to `TIFFRGBAImageOK@LIBTIFF_4.0'
../../lib/librender.so: undefined reference to `TIFFWriteRawStrip@LIBTIFF_4.0'
../../lib/librender.so: undefined reference to `TIFFReadRGBAImage@LIBTIFF_4.0'
../../lib/librender.so: undefined reference to `TIFFSetField@LIBTIFF_4.0'
../../lib/librender.so: undefined reference to `TIFFSetErrorHandler@LIBTIFF_4.0'
../../lib/librender.so: undefined reference to `_TIFFmalloc@LIBTIFF_4.0'
../../lib/librender.so: undefined reference to `TIFFSetDirectory@LIBTIFF_4.0'
../../lib/librender.so: undefined reference to `TIFFReadScanline@LIBTIFF_4.0'
collect2: error: ld returned 1 exit status
make[2]: *** [bin/vapor] Error 1
make[1]: *** [apps/vaporgui/CMakeFiles/vapor.dir/all] Error 2
make: *** [all] Error 2
```

--- a/lib/flow/Advection.cpp
+++ b/lib/flow/Advection.cpp
@@ -90,6 +90,7 @@ int Advection::AdvectSteps(Field *velocity, double deltaT, size_t maxSteps, ADVE
             }
 
             if (rv == SUCCESS) {
+                // Bookmark_1
                 // The new particle *may* be the same as the old particle in case
                 // there's a sink, meaning the velocity is zero.
                 // In that case, we mark p1 as "special" and terminate the current stream.
@@ -104,6 +105,7 @@ int Advection::AdvectSteps(Field *velocity, double deltaT, size_t maxSteps, ADVE
                     numberOfSteps++;
                 }
             } else if (rv == MISSING_VAL) {
+                // Bookmark_2
                 // This is the annoying part: there are multiple possiblities.
                 // 1) past0 is really located at a missing value location;
                 // 2) past0 is inside the volume, but really close to the boundary,
@@ -198,6 +200,9 @@ int Advection::AdvectTillTime(Field *velocity, double startT, double deltaT, dou
         if (p0.time < startT)      // Skip this stream if it didn't advance to startT
             continue;
 
+        if (p0.IsSpecial())
+            continue;
+
         size_t thisStep = 0;
 
         while (p0.time < targetT) {
@@ -229,7 +234,6 @@ int Advection::AdvectTillTime(Field *velocity, double startT, double deltaT, dou
                 } else {
                     break;    // break the while loop
                 }
-
             } // Finish the out-of-volume condition
 
             double dt = deltaT;
@@ -258,19 +262,68 @@ int Advection::AdvectTillTime(Field *velocity, double startT, double deltaT, dou
                 _printNonZero(rv, __FILE__, __func__, __LINE__);
                 break;
             }
-            if (rv != 0) {    // Advection wasn't successful for some reason...
-                break;
-            } else {    // Advection successful, keep the new particle.
-                happened = true;
-                s.push_back(p1);
-                p0 = p1;
-            }
+
+            if (rv == SUCCESS) {
+                // Check out Bookmark_1
+                if (p1.location == p0.location) {
+                    p1.SetSpecial(true);
+                    s.push_back(p1);
+                    _separatorCount[streamIdx]++;
+                    break;
+                } else {
+                    happened = true;
+                    s.push_back(p1);
+                    p0 = p1;
+                }
+            } else if (rv == MISSING_VAL) {
+                // Check out Bookmark_2
+                glm::vec3 vel;
+                bool isMissing = (velocity->GetVelocity(p0.time, p0.location, vel) == MISSING_VAL);
+                bool isInside = velocity->InsideVolumeVelocity(p0.time, p0.location);
+
+                if (isInside && isMissing) {
+                    p1.SetSpecial(true);
+                    s.push_back(p1);
+                    _separatorCount[streamIdx]++;
+                    break;
+                } else if (isInside && (!isMissing)) {
+                    rv = _advectEuler(velocity, p0, dt, p1);
+                    assert(rv == 0);
+                    s.push_back(p1);
+                    p0 = p1;
+                } else {
+                    auto loc = p0.location;
+                    for (int i = 0; i < 3; i++) {
+                        if (_isPeriodic[i])
+                            loc[i] = _applyPeriodic(loc[i], _periodicBounds[i][0], _periodicBounds[i][1]);
+                    }
+
+                    if (velocity->InsideVolumeVelocity(p0.time, loc)) {
+                        p1.SetSpecial(true);
+                        auto it = s.end();
+                        --it;
+                        s.insert(it, p1);
+                        it = s.end();
+                        --it;
+                        it->location = loc;
+                        _separatorCount[streamIdx]++;
+                    } else {
+                        p1.SetSpecial(true);
+                        s.push_back(p1);
+                        _separatorCount[streamIdx]++;
+                        break;
+                    }
+                }
+            } // finish handling missing value
 
             // Another termination criterion: when advecting at least 10,000 steps and
             // more than 10X more than the previous max num of steps.
             //
             if (++thisStep == maxSteps) {
                 thisStep = maxSteps / 10;
+                p1.SetSpecial(true);
+                s.push_back(p1);
+                _separatorCount[streamIdx]++;
                 break;
             }
         }    // Finish advecting one particle

--- a/lib/flow/Advection.cpp
+++ b/lib/flow/Advection.cpp
@@ -200,12 +200,13 @@ int Advection::AdvectTillTime(Field *velocity, double startT, double deltaT, dou
         if (p0.time < startT)      // Skip this stream if it didn't advance to startT
             continue;
 
-        if (p0.IsSpecial())
+        if (p0.IsSpecial())       // Also skip this tream if it was marked special.
             continue;
 
         size_t thisStep = 0;
 
         while (p0.time < targetT) {
+
             // Check if the particle is inside of the volume.
             // Wrap it along periodic dimensions if applicable.
             if (!velocity->InsideVolumeVelocity(p0.time, p0.location)) {
@@ -219,8 +220,13 @@ int Advection::AdvectTillTime(Field *velocity, double startT, double deltaT, dou
                         locChanged = true;
                     }
                 }
-                if (!locChanged)    // no dimension is periodic
+                if (!locChanged) {  // no dimension is periodic, append a separator
+                    Particle separator;
+                    separator.SetSpecial(true);
+                    s.push_back(separator);
+                    _separatorCount[streamIdx]++;
                     break;          // break the while loop
+                }
 
                 // See if the new location is inside of the volume
                 if (velocity->InsideVolumeVelocity(itr->time, loc)) {
@@ -231,7 +237,11 @@ int Advection::AdvectTillTime(Field *velocity, double startT, double deltaT, dou
                     separator.SetSpecial(true);
                     s.insert(itr, separator);
                     _separatorCount[streamIdx]++;
-                } else {
+                } else {  // Still outside, so we terminate the stream!
+                    Particle separator;
+                    separator.SetSpecial(true);
+                    s.push_back(separator);
+                    _separatorCount[streamIdx]++;
                     break;    // break the while loop
                 }
             } // Finish the out-of-volume condition

